blockstore/apps/bundles/models.py

```diff
@@ -0,0 +1,223 @@
+    the bundle can be modified without breaking anything.
+
+    Target Scale: 100M rows
+    """
+    id = models.BigAutoField(primary_key=True)
+    uuid = models.UUIDField(default=uuid.uuid4, unique=True)
+    title = models.CharField(max_length=MAX_CHAR_FIELD_LENGTH, db_index=True)
+
+    # Comment for now just to make testing easier.
+    # collection = models.ForeignKey(
+    #     Collection, related_name="bundles", related_query_name="bundle"
+    # )
+
+    slug = models.SlugField(allow_unicode=True)  # For pretty URLs
+    description = models.TextField(max_length=10_000)
+
+    def __str__(self):
+        return "Bundle {} - {}".format(self.uuid, self.slug)
+
+
+@attr.s(frozen=True)
+class BundleVersionKey():
+    """
+    Do we really want this class at all?
+
+    BundleVersion ID serialization could be left to the view layer, since we're
+    not storing serialized BundleVersion IDs anywhere as one thing. It's mostly
+    just input shorthand.
+    """
+    bundle_uuid = attr.ib(type=uuid.UUID)
+    version_num = attr.ib(type=int)
+
+    @classmethod
+    def parse(cls, key_str):
+        """
+        Format for this is {UUID}@{version_num}.
+        """
+        try:
+            bundle_uuid_str, version_str = key_str.split('@')
+            bundle_uuid = uuid.UUID(hex=bundle_uuid_str)
+            version_num = int(version_str)
+        except Exception:
+            raise ValueError("Could not parse BundleVersionKey from {}".format(part_id_str))
+
+        return cls(bundle_uuid, version_num)
+
+    def __str__(self):
+        return "{}@{}".format(self.bundle_uuid, self.version_num)
+
+
+class BundleVersion(models.Model):
+    """
+    The contents of a BundleVersion are immutable (the snapshot it points to),
+    but the metadata about a BundleVersion (e.g. change_description) can be
+    changed. Other entities in the system that need to attach metadata to
+    versions of Bundles should use this model and not reference BundleSnapshots
+    directly.
+
+    Target Scale: 1B rows
+    """
+    id = models.BigAutoField(primary_key=True)
+    bundle = models.ForeignKey(
+        Bundle, related_name="versions", related_query_name="version"
+    )
+    version_num = models.PositiveIntegerField()
+
+    snapshot_digest = models.BinaryField(max_length=20, db_index=True)
+    # compatibility_hash = Let content determine a hash so that two versions can
+    #                      be compared for compatibility (ORA2 does this). A
+    #                      proxy for breaking change or major/minor versions?
+
+    change_description = models.TextField(max_length=1_000, blank=True)
+
+    @staticmethod
+    @receiver(snapshot_created)
+    def listen_for_snapshot_creation(**kwargs):
+        """
+        The BundleStore layer doesn't know we exist, but it emits signals when a
+        new BundleSnapshot is created. We listen for that signal and create a
+        new BundleVersion that points to it.
```

Comment:
> Is the Blockstore supposed to be sufficient as the primary storage for authoring interfaces?

Yes.

> 1. If yes, should every save-preview cycle result in a new version? Or if the author uploads a few files one by one to a bundle in an authoring session, should each file addition result in a new version?

Absolutely not. We've walked down that path with SplitMongo, and I have no desire to go down it again. I finally wrote a [script](https://github.com/edx/tubular/blob/master/tubular/scripts/structures.py) to clean up most of that garbage a year ago. There's no value in storing all of those as separate versions, and so we shouldn't.

I think there should be a draft directory where you read and write to your hearts content and nothing is committed to a snapshot or version. At some point, you decide, "yes, this is a new version right here", and then you press a button and it becomes a version. I'm not sure the exact form that will take in implementation -- do we copy everything from the most recent version when starting a draft? Lazy copy-on-write? What happens if a person creates a new version while we're still editing our draft? But yeah, definitely not thousands of useless versions. Ugh.

> 2. What purpose do version numbers serve in an authoring interface?

It's like the basic versioning information you can add in wikis (but nobody does in practice). They can associate messages describing the update, and they'd get a list of versions, dates, who did the editing, and commit messages (I realize most of those aren't there yet).

> Is the author going to get a long list of numbers or are there going to be any other annotations like tags?

The expectation would be that other things annotate BundleVersions, and that would include search and tagging.

> If the tags are going to be stored in the tagging service for example, will it possible for authors to use them for linking from one bundle to another or will they have to find out the corresponding version number and use that? Or should version tags be first class entities in the Blockstore?

Sorry, there are a few different types of tagging that have been floating around in different conversations and I'm not sure what you mean by that.

> 3. Can we add a created_at field? I think that is going to be very helpful.

Yeah. I punted on it for a moment because we weren't using it in a view yet and I was punting on whether to add the drf-extensions dependency or just use straight timestamps, but either way they should be in there. I'll add them.

> Is there any document which has a list of authoring workflows from the product perspective so we can see if/how the APIs will work in such scenarios?

@bradenmacdonald would likely have the latest and greatest on that front.

Reply:
> Is there any document which has a list of authoring workflows from the product perspective so we can see if/how the APIs will work in such scenarios?

> @bradenmacdonald would likely have the latest and greatest on that front.

There's a WIP version out there but I haven't got access to it yet. Will share with y'all as soon as it's ready. The LX team has been making some major revisions.

Reply:
> Absolutely not. We've walked down that path with SplitMongo, and I have no desire to go down it again. I finally wrote a script to clean up most of that garbage a year ago. There's no value in storing all of those as separate versions, and so we shouldn't.

Since you had added extensive comments about a lot of other things in the code, but nothing here, I was a bit unsure. * phew * ;) This sounds good. 

I have created a ticket so we can tackle this separately: https://tasks.opencraft.com/browse/LX-50

> It's like the basic versioning information you can add in wikis (but nobody does in practice). They can associate messages describing the update, and they'd get a list of versions, dates, who did the editing, and commit messages (I realize most of those aren't there yet).
> Sorry, there are a few different types of tagging that have been floating around in different conversations and I'm not sure what you mean by that.

I meant something like git tags that are used for referring to specific versions and can have some meaning. What I am getting is that if someone wants to add semantic versions  e.g: release-2018-01-01 or v2.3.3, they should put those in the description field and links between bundles must use the integer `version_num` for specifying the `BundleVersion`. Is that correct?